### ISSUE TITLE
Detailed knex/driver errors passed through in production

### ISSUE THREAD

**Issue by nickschot** [postlight/lux#697]
*   **Platform**: macOS 10.12
*   **Database**: pg/postgres 9.6
*   **Lux Version**: 1.1.7
*   **Node Version**: 7.9.0

When saving a model via lux which does not adhere to the database constraints, knex/pg might throw an error like below:

`detail: "update "users" set "email" = $1, "updated_at" = $2 where "id" = $3 - duplicate key value violates unique constraint "users_email_unique""` together with a `500 Internal Server Error`

In the current situation these detailed messages are also passed through in the response in production mode. This is unwanted behaviour.

**Comment by nickschot** [comment]
What is the acceptable solution here? Removing the "detail" for 500 messages, or would there be more cases where detail shouldn't be present in production?

**Comment by jamemackson** [comment]
I'd say this, ideally IMO, should be configurable similar to a log level.  

A `high` level might return what is returned today, 
A `medium` might just omit the "detail" field and 
A `low` might omit both the detail and the title fields only return ing a status
A `secure` might just default to a blanket 'unauthorized' message or something...

Just some thoughts...

**Comment by nickschot** [comment]
It should indeed be coupled to the log level. Why would you wan't to omit title/status though? That's the only way to convey what went wrong or right.

**Comment by jamemackson** [comment]
@nickschot - To be clear, I don't think it should be 'coupled to' but configured 'similarly to' the log level.  I would likely want very different internal logging settings vs how verbose I want external errors to be.

> Why would you want to omit title/status though?

Depending on how secure I want my application to be would, in my mind, dictate how little detail I want to return to a user (or potential hacker) of my api when an error occurs.  If I have a very sensitive set of data I might not want to return any clues for a potential evil doer that might help clue them into how they can adjust their attack vector.   it would also likely depend on if they were authenticated or not.  I would feel more ok with returning more details to an authenticated request vs an unauthenticated one.  

Some of this is obviously an implementation detail in how I setup my authentication of the api but I think it does still relate here.

**Comment by zacharygolba** [comment]
Just adding my $0.02 here. I think that log levels and error responses shouldn't be coupled together in any way. IMO, you shouldn't be able to configure error responses. It's never a good idea to leak SQL statements in error messages in production.

**tldr;**

You shouldn't have to think about the security risks of error messages beyond making sure your production application is running in production mode.

> What is the acceptable solution here? Removing the "detail" for 500 messages, or would there be more cases where detail shouldn't be present in production?

I think a solid solution for common errors would be mapping knex error types to strings for substitution with the original error message and/or stack trace when running in production mode. There could also be a "catch all"  default error message (something like "An unknown error occurred.") if a mapping is not found.
